### PR TITLE
IGNITE-17312 Java thin: Implement unified exception handling

### PR DESCRIPTION
**Pull request by ptupitsyn**
* Remove `ClientErrorCode`. Use `IgniteException#code` instead.
* Add `Client` error group.
* Send all details to reconstruct the exception on the client side:
  * code
  * traceId
  * class name
  * message
  * stack trace (when `sendServerExceptionStackTraceToClient` config flag is true)
* Fix exception propagation and unwrapping in client logic.
* Copy `ExceptionUtils` from `X` class in Ignite 2 to get stack trace as string.
* Fix .NET part in a minimal way, proper update to be done in IGNITE-17390.

### CODE REVIEW

File: modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java [hunk 1]
```diff
@@ -296,9 +297,9 @@ private <T> void handleServiceAsync(final CompletableFuture<T> fut,
      * @return host:port_range address lines parsed as {@link InetSocketAddress} as a key. Value is the amount of appearences of an address
      *      in {@code addrs} parameter.
      */
-    private static Map<InetSocketAddress, Integer> parsedAddresses(String[] addrs) throws IgniteClientException {
+    private static Map<InetSocketAddress, Integer> parsedAddresses(String[] addrs) {
         if (addrs == null || addrs.length == 0) {
-            throw new IgniteClientException("Empty addresses");
+            throw new IgniteException(UNKNOWN_ERR, "Empty addresses");
```
Comment (isapego) [inline modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java:302]: Maybe something like `USER_INPUT_ERR`?

File: modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java [hunk 2]
```diff
@@ -557,7 +559,7 @@ private ClientChannel getDefaultChannel() {
             }
         }
 
-        throw new IgniteClientConnectionException("Failed to connect", failure);
+        throw new IgniteClientConnectionException(UNKNOWN_ERR, "Failed to connect", failure);
```
Comment (isapego) [inline modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java:562]: Why not `CONNECTION_ERR`?

File: modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java [hunk 3]
```diff
@@ -683,7 +683,8 @@ private ClientChannel getOrCreateChannel(boolean ignoreThrottling)
                     }
 
                     if (!ignoreThrottling && applyReconnectionThrottling()) {
-                        throw new IgniteClientConnectionException("Reconnect is not allowed due to applied throttling");
+                        //noinspection NonPrivateFieldAccessedInSynchronizedContext
+                        throw new IgniteClientConnectionException(UNKNOWN_ERR, "Reconnect is not allowed due to applied throttling");
```
Comment (isapego) [inline modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java:687]: `CONNECTION_ERR`?

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java [hunk 4]
```diff
@@ -291,30 +271,58 @@ private void processNextMessage(ByteBuf buf) throws IgniteClientException {
...
     }
 
+    /**
+     * Unpacks request error.
+     *
+     * @param unpacker Unpacker.
+     * @return Exception.
+     */
+    private IgniteException readError(ClientMessageUnpacker unpacker) {
+        var traceId = unpacker.tryUnpackNil() ? UUID.randomUUID() : unpacker.unpackUuid();
```
Comment (isapego) [inline modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:303]: Is random UUID a proper way to deal with this situation? Won't it cause any issues? Did you discuss it with guys who designed it?

File: modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientAsyncResultSet.java [hunk 5]
```diff
@@ -134,11 +136,11 @@ public CompletionStage<? extends AsyncResultSet> fetchNextPage() {
         requireResultSet();
 
         if (closed) {
-            return CompletableFuture.failedFuture(new IgniteClientException("Cursor is closed."));
+            return CompletableFuture.failedFuture(new IgniteException(UNKNOWN_ERR, "Cursor is closed."));
         }
 
         if (!hasMorePages()) {
-            return CompletableFuture.failedFuture(new IgniteClientException("No more pages."));
+            return CompletableFuture.failedFuture(new IgniteException(UNKNOWN_ERR, "No more pages."));
```
Comment (isapego) [inline modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientAsyncResultSet.java:143]: Do we use `UNKNOWN_ERR` as generic error code?

File: modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java [hunk 6]
```diff
@@ -177,21 +181,26 @@ void testHandshakeInvalidVersionReturnsError() throws Exception {
             // Read response.
             var unpacker = MessagePack.newDefaultUnpacker(sock.getInputStream());
             var magic = unpacker.readPayload(4);
-            unpacker.skipValue(3);
-            var len = unpacker.unpackInt();
-            var major = unpacker.unpackInt();
+            unpacker.readPayload(4); // Length.
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/integrationTest/java/org/apache/ignite/client/handler/ItClientHandlerTest.java:184]: Message length is a raw int32 value, not a MessagePack type.

File: modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java [hunk 7]
```diff
@@ -191,7 +194,7 @@ private void handshake(ChannelHandlerContext ctx, ClientMessageUnpacker unpacker
             var clientVer = ProtocolVersion.unpack(unpacker);
 
             if (!clientVer.equals(ProtocolVersion.LATEST_VER)) {
-                throw new IgniteException("Unsupported version: "
+                throw new IgniteException(PROTOCOL_ERR, "Unsupported version: "
```
[the remark this answers is not in the document]
Comment (ptupitsyn) [reply modules/client-handler/src/main/java/org/apache/ignite/client/handler/ClientInboundMessageHandler.java:197]: Good point, added PROTOCOL_COMPATIBILITY_ERR.

File: modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java (hunk 1 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java:302]: Added `CONFIGURATION_ERR`.

File: modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java (hunk 2 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java:562]: Agree, fixed.

File: modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java (hunk 3 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/ReliableChannel.java:687]: Fixed.

File: modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java (hunk 4 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/TcpClientChannel.java:303]: This code is not correct - the server always sends `traceId` and `code`, no need for nil checks here. Fixed.

File: modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientAsyncResultSet.java (hunk 5 above)
Comment (ptupitsyn) [reply modules/client/src/main/java/org/apache/ignite/internal/client/sql/ClientAsyncResultSet.java:143]: In most cases we should not. I've added TODOs here with a link to [IGNITE-17135 SQL API: Exceptions are different on server and client](https://issues.apache.org/jira/browse/IGNITE-17135).

**Review by isapego — APPROVED** [review]

File: modules/api/src/main/java/org/apache/ignite/lang/IgniteException.java [hunk 8]
```diff
@@ -252,4 +254,30 @@ public int errorCode() {
...
+     * Wraps another exception in IgniteException, extracting {@link #traceId} and {@link #code} when the specified exception
+     * or one of its causes is an IgniteException itself.
+     *
+     * @param e Internal exception.
+     * @return Public exception.
+     */
+    public static IgniteException wrap(Throwable e) {
+        e = ExceptionUtils.unwrapCause(e);
+
+        if (e instanceof IgniteException) {
```
Comment (sk0x50) [inline modules/api/src/main/java/org/apache/ignite/lang/IgniteException.java:269]: It seems to me that we need to add support for `IgniteCheckedException` as well.

Comment (ptupitsyn) [reply modules/api/src/main/java/org/apache/ignite/lang/IgniteException.java:272]: Good catch, done.

**Pull request merged**